lib/hanami/application.rb

```diff
@@ -1,203 +1,84 @@
-      end
-
-      def prepared?
-        !!@prepared
-      end
-
-      def booted?
-        !!@booted
-      end
-
-      def router
-        raise "Application not yet prepared" unless prepared?
-
-        @_mutex.synchronize do
-          @_router ||= load_router
-        end
-      end
-
-      def rack_app
-        @rack_app ||= router.to_rack_app
-      end
-
-      def slices
-        @slices ||= SliceRegistrar.new(self)
-      end
-
-      def register_slice(...)
-        slices.register(...)
-      end
-
-      def register(...)
-        container.register(...)
-      end
-
-      def register_provider(...)
-        container.register_provider(...)
-      end
-
-      def start(...)
-        container.start(...)
-      end
-
-      def key?(...)
-        container.key?(...)
-      end
+      attr_reader :autoloader, :configuration
 
-      def keys
-        container.keys
-      end
-
-      def [](...)
-        container.[](...)
-      end
-
-      def resolve(...)
-        container.resolve(...)
-      end
-
-      def settings
-        @_settings ||= load_settings
-      end
-
-      def namespace
-        application_name.namespace
-      end
-
-      def root
-        configuration.root
-      end
-
-      def inflector
-        configuration.inflector
+      def application_name
+        slice_name
       end
 
       private
 
       def prepare_base_load_path
-        base_path = File.join(root, LIB_DIR)
-        $LOAD_PATH.unshift base_path unless $LOAD_PATH.include?(base_path)
+        base_path = root.join(LIB_DIR)
+        $LOAD_PATH.unshift(base_path) unless $LOAD_PATH.include?(base_path)
       end
 
       def prepare_all
-        load_settings
-        prepare_container_plugins
-        prepare_container_base_config
-        prepare_container_component_dirs
-        prepare_container_consts
-        container.configured!
-        prepare_slices
-        # For the application, the autoloader must be prepared after the slices, since
-        # they'll be configuring the autoloader with their own dirs
-        prepare_autoloader
-      end
-
-      def prepare_container_plugins
-        container.use(:env, inferrer: -> { Hanami.env })
-        container.use(:zeitwerk, loader: autoloader, run_setup: false, eager_load: false)
+        # Make app-wide notifications available as early as possible
         container.use(:notifications)
-      end
 
-      def prepare_container_base_config
-        container.config.root = configuration.root
-        container.config.inflector = configuration.inflector
+        # Ensure all basic slice preparation is complete before we make adjustments below
+        # (which rely on the basic prepare steps having already run)
+        super
+
+        # Run specific prepare steps for the application slice. Note also that some
+        # standard steps have been skipped via the empty method overrides below.
+        prepare_application_component_dirs
+        prepare_application_providers
```

Comment:
I just want to put on the record that I'm not particularly enamoured with this solution here, but I think it's good enough for now.

Longer term I could see the slice having a registered list of setup steps (perhaps one that can be added to by extension gems), and at that point, the app class could explicitly delete the unwanted steps, rather than blanking them out like this.